### PR TITLE
Merge layers

### PR DESCRIPTION
**Pull request by cirelli94**
I added a feature to obtain a single `map[string]interface{}`, instead of the usual slice, containing all the layers.
And I also added a method that will use this merge: a MergeAndDecode method that will decode our configuration directly in a struct using mapstructure, that is already a dependency of this project.

I hope you'll be glad of my adds and you will merge this feature to your project! We would like to use onion in our, I liked the layer concept and how you have implemented it.

Thanks in advance

### CODE REVIEW

**Comment by fzerorubigd** [comment]
There is similar functionality in the file writer (https://github.com/goraz/onion/blob/master/onionwriter/onion_writer.go#L13) 

But I prefer your implementation. The only problem is the new dependency in the main package. 
My suggestion is to move this code to the onionwriter, refactor `SerializeOnion` to use your method, export two functions: `MergedLayersData(o *Onion) map[string]interface{}` and also `MergeAndDecode(o *Onion, obj interface{}) error`

**Comment by fzerorubigd** [comment]
Also just an idea, with your function we can have a layer that actually is another onion object. I am not sure if it is useful or not tho.

**Comment by cirelli94** [comment]
> There is similar functionality in the file writer (https://github.com/goraz/onion/blob/master/onionwriter/onion_writer.go#L13)
> 
> But I prefer your implementation. The only problem is the new dependency in the main package.
> My suggestion is to move this code to the onionwriter, refactor `SerializeOnion` to use your method, export two functions: `MergedLayersData(o *Onion) map[string]interface{}` and also `MergeAndDecode(o *Onion, obj interface{}) error`
> 
> Also please add tests for these new functions.

I would like to move my code inside onionwriter but there's a problem if I call onionwriter from onion:
```
can't load package: import cycle not allowed
package github.com/goraz/onion
        imports github.com/goraz/onion/onionwriter
        imports github.com/goraz/onion
```

> Also just an idea, with your function we can have a layer that actually is another onion object. I am not sure if it is useful or not tho.

I think I'll use this function for implement the folder feature.

**Comment by fzerorubigd** [comment]
I can not understand why you import `configwriter` in `onion`?

**Comment by cirelli94** [comment]
> I can not understand why you import `configwriter` in `onion`?

`onionwriter`! If I move my code that do the merge in `onionwriter`, then I need to recall it from onion.go for the apis 
```
MergedLayersData(o *Onion) map[string]interface{}
MergeAndDecode(o *Onion, obj interface{}) error
```

**Comment by fzerorubigd** [comment]
Sure, it is already imported in the onionwriter, my question was why you need to import onionwriter in onion. (the circular import)

**Comment by cirelli94** [comment]
> Sure, it is already imported in the onionwriter, my question was why you need to import onionwriter in onion. (the circular import)

I wanted to move a merge method in onionwriter.
Ok, I'm going to finish to write even the folder part and commit, then you will tell me :)

**Comment by fzerorubigd** [comment]
Cool! Thank you :)

**Comment by cirelli94** [comment]
I think I did what you asked me!
I had to create a package for a function that is used both from onionwriter and onion:  `utils` package...
I'm going to write some test for folder-layer and that's all folks!

**Comment by fzerorubigd** [comment]
Just a minor issue: why the directory_layer is not an external layer inside the `layers` folder (https://github.com/goraz/onion/tree/master/layers)

**Comment by cirelli94** [comment]
I didn't understand why some layers are outside and some inside layers: however I now moved as you asked.
I have some problems with the test with go versions before 1.14, I don't know why...

**Comment by fzerorubigd** [comment]
This is why onion is different. the problem I had was having too many
dependencies for one simple task. If you don't want to use the etcdlayer,
adding etcd library to your code is not a good idea. also the same for the
watch file layer and in this case, the folder layer. in the end, the
compiler merges them into one single file, so there is no performance
problem with many files. (like python/PHP and scripting language in
general)
So basically you can add any layer, without modifying the main library (the
onion package itself).

File: onion_test.go [hunk 1]
```diff
@@ -182,6 +182,8 @@ func TestOnion(t *testing.T) {
 		o1.AddLayers(lm)
 		So(o1.GetIntDefault("test0", 0), ShouldEqual, 1)
 	})
+
+
```
Comment (fzerorubigd) [inline onion_test.go]: please rest this file to its original state.

**Comment by cirelli94** [comment]
I also misnamed the directorylayer package.
I fixed the empty lines you asked and something more in directorylayer.
Maybe maybe it's the right time...

**Comment by cirelli94** [comment]
Tests are going wrong because behaviour of https://golang.org/pkg/io/ioutil/#TempDir has changed.
